src/Microsoft.AspNet.Server.Kestrel/Http/Connection.cs

```diff
@@ -63,13 +65,53 @@ public Connection(ListenerContext context, UvStreamHandle socket) : base(context
             ConnectionControl = this;
         }
 
+        private void ExtractIP()
+        {
+            // storage for binary IP addresses
+            Libuv.sockaddr peer;
+            Libuv.sockaddr local;
+            // length of sockaddr structs
+            var peerLength = Marshal.SizeOf(typeof(Libuv.sockaddr));
+            var localLength = peerLength;
+            // buffers for addresses
+            var peerString = new StringBuilder(50);
+            var localString = new StringBuilder(50);
+            // storage for errors during decoding processes
+            Exception peerException;
+            Exception localException;
+
+            // get local and peer IPs in binary form
+            _socket.Libuv.tcp_getsockname((UvTcpHandle)_socket, out local, ref localLength);
+            _socket.Libuv.tcp_getpeername((UvTcpHandle)_socket, out peer, ref peerLength);
+            // decode as IPv4 addresses
+            _socket.Libuv.ip4_name(ref local, localString, 50, out localException);
```

Comment:
The connection is built from a listening socket and this is either IPv4 or IPv6. Why not cache that information and call the appropriate function based on that?

Side note: In my patched version of Kestrel, I'm using a listening socket on `IPAdress.IPv6Any` instead of `IPAddress.Any`, which is a dual-stack socket per default. In that scenario, using the IPv6 version is still the correct option and IPv4-adresses are formatted like `::ffff:192.168.1.2` as expected. I haven't found anything in Kestrel that's not ready for IPv6, libuv [certainly is](http://docs.libuv.org/en/latest/migration_010_100.html#unification-of-ipv4-ipv6-apis).